### PR TITLE
SNOW-1418523: concurrent file operations

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1418523

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   This change adds tests for ensuring fileIO works correctly in multithreading mode. Since file io relies on cursor, make cursor safe in https://github.com/snowflakedb/snowpark-python/pull/2282 is sufficient to achieve this goal.

### CODE REVIEW

File: tests/integ/test_multithreading.py [hunk 1]
```diff
@@ -122,3 +125,67 @@ def test_action_ids_are_unique(session):
...
+            # assert two files are identical
+            with open(upload_file_path, "rb") as upload_fd, open(
+                download_file_path, "rb"
+            ) as download_fd:
+                assert get_file_hash(upload_fd) == get_file_hash(download_fd)
+
+    with tempfile.TemporaryDirectory() as download_dir:
+        with ThreadPoolExecutor(max_workers=10) as executor:
+            for file_path in resources_files:
+                executor.submit(put_and_get_file, file_path, download_dir)
```
Comment (sfc-gh-aling) [inline tests/integ/test_multithreading.py:191]: shall we verify all the files are downloaded in the end of the test?

Comment (sfc-gh-aalam) [reply tests/integ/test_multithreading.py:191]: we do that check. see lines 178 - 186. Do you want me to add some more checks outside of this?

Comment (sfc-gh-aling) [reply tests/integ/test_multithreading.py:191]: emmm that one is verifying a single file right?
yeah, I'm thinking outside the loop
checking all files show up in the download dir like

os.list(downdir_dir) == [
        test_files.test_file_csv,
        test_files.test_file2_csv,
        test_files.test_file_json,
        test_files.test_file_csv_header,
        test_files.test_file_csv_colon,
        test_files.test_file_csv_quotes,
        test_files.test_file_csv_special_format,
        test_files.test_file_json_special_format,
        test_files.test_file_csv_quotes_special,
        test_files.test_concat_file1_csv,
        test_files.test_concat_file2_csv
]

Comment (sfc-gh-aalam) [reply tests/integ/test_multithreading.py:191]: sure. I can add that check as well.

Comment (sfc-gh-aalam) [reply tests/integ/test_multithreading.py:191]: added

Comment (sfc-gh-aling) [reply tests/integ/test_multithreading.py:191]: thanks!

**Review by sfc-gh-aling — APPROVED** [review]

**Review by sfc-gh-yuwang — APPROVED** [review]

**Pull request merged**